src/local/room_position.rs

```diff
@@ -0,0 +1,375 @@
+///
+///   let pos = LocalRoomPosition::new(20, 21, "E5N6".parse().unwrap());
+///   let result = js! {
+///       let pos = pos_from_packed(@{pos.packed_repr()});
+///       pos.roomName
+///   };
+///   ```
+///
+/// [`bincode`]: https://github.com/servo/bincode
+/// [`RoomPosition::local`]: crate::RoomPosition::local
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct LocalRoomPosition {
+    /// A bit-packed integer, containing, from highest-order to lowest:
+    ///
+    /// - 1 byte: (room_x) + 128
+    /// - 1 byte: (room_y) + 128
+    /// - 1 byte: x
+    /// - 1 byte: y
+    ///
+    /// For `Wxx` rooms, `room_x = -xx - 1`. For `Exx` rooms, `room_x = xx`.
+    ///
+    /// For `Nyy` rooms, `room_y = -yy - 1`. For `Syy` rooms, `room_y = yy`.
+    ///
+    /// This is the same representation used in the Screeps server, allowing for
+    /// easy translation. Besides the method names and display representation,
+    /// this is the one part of LocalRoomPosition copied directly from the
+    /// engine code.
+    packed: u32,
+}
+
+impl fmt::Debug for LocalRoomPosition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LocalRoomPosition")
+            .field("packed", &self.packed)
+            .field("x", &self.x())
+            .field("y", &self.y())
+            .field("room_name", &self.room_name())
+            .finish()
+    }
+}
+
+impl fmt::Display for LocalRoomPosition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[room {} pos {},{}]",
+            self.room_name(),
+            self.x(),
+            self.y()
+        )
+    }
+}
+
+impl LocalRoomPosition {
+    /// Create a new LocalRoomPosition
+    ///
+    /// # Panics
+    ///
+    /// Will panic if either `x` or `y` is larger than 49, or if `room_name` is
+    /// outside of the range `E127N127 - W127S127`.
+    #[inline]
+    pub fn new(x: u32, y: u32, room_name: LocalRoomName) -> Self {
+        assert!(x < 50, "out of bounds x: {}", x);
+        assert!(y < 50, "out of bounds y: {}", y);
+        assert!(
+            (-128..=127).contains(&room_name.x_coord),
+            "out of bounds room_x: {}",
+            room_name.x_coord,
+        );
+        assert!(
+            (-128..=127).contains(&room_name.y_coord),
+            "out of bounds room_y: {}",
+            room_name.y_coord,
+        );
+        let room_x = (room_name.x_coord + HALF_WORLD_SIZE) as u32;
+        let room_y = (room_name.y_coord + HALF_WORLD_SIZE) as u32;
+
+        LocalRoomPosition {
+            packed: (room_x << 24) | (room_y << 16) | (x << 8) | y,
+        }
+    }
+
+    #[inline]
+    pub fn packed_repr(self) -> i32 {
+        self.packed as i32
+    }
+
+    #[inline]
+    pub fn from_packed(packed: i32) -> Self {
+        LocalRoomPosition {
+            packed: packed as u32,
+        }
+    }
+
+    #[inline]
+    fn room_x(self) -> i32 {
```

Comment:
Without looking at the code, it's unclear whether `room_x` is the x coordinates of the room, or the x coordinate within a room.

Reply:
Documentation added!